torch/csrc/jit/graph_executor.cpp

```diff
@@ -0,0 +1,396 @@
+#include "torch/csrc/jit/passes/common_subexpression_elimination.h"
+#include "torch/csrc/jit/passes/peephole.h"
+#include "torch/csrc/jit/passes/graph_fuser.h"
+#include "torch/csrc/jit/passes/inplace_check.h"
+#include "torch/csrc/jit/passes/batch_mm.h"
+
+#include "torch/csrc/autograd/function.h"
+
+#include <unordered_map>
+
+namespace torch { namespace jit {
+
+namespace {
+
+using tensor_list = std::vector<at::Tensor>;
+using variable_tensor_list = tensor_list;
+using Variable = autograd::Variable;
+using autograd::variable_list;
+
+struct ExecutionPlanAutogradFunction : public autograd::Function {
+  ExecutionPlanAutogradFunction(GraphExecutor graph, size_t capture_size)
+  : graph(std::move(graph)), captures(capture_size) {}
+  virtual variable_list apply(const variable_list& inputs) override {
+    // TODO: expensive copies here to convert to/from tensor_list
+    // TODO: becuse inputs is passed by const reference there is no
+    // way to release tensors incrementally as this runs
+    variable_tensor_list all_inputs;
+    all_inputs.reserve(captures.size() + inputs.size());
+    for(auto & sv : captures) {
+      all_inputs.push_back(sv.unpack(this->shared_from_this()));
+    }
+    all_inputs.insert(all_inputs.end(), inputs.begin(), inputs.end());
+    auto tensors = graph.run(std::move(all_inputs));
+    // TODO: another copy that needs to be removed
+    return autograd::variable_list(tensors.begin(), tensors.end());
+  }
+private:
+  friend struct ExecutionPlan;
+  GraphExecutor graph;
+  std::vector<autograd::SavedVariable> captures;
+};
+
+
+// an optimized way of executing the subgraph computed directly on
+// tensors rather than Variables.
+// This will unwrap Variables, run the plan, and re-wrap them.
+// It can optionally also have a gradient which is hooked up
+// to the output Variables if present.
+struct ExecutionPlan {
+  ExecutionPlan(std::shared_ptr<Graph> & graph)
+  : f(graph) {}
+  ExecutionPlan(std::shared_ptr<Graph> & graph, Gradient grad)
+  : f(graph), grad(std::move(grad)), grad_executor(this->grad.df) {}
+
+  variable_tensor_list run(variable_tensor_list inputs) {
+    if(grad) {
+      return runWithGrad(std::move(inputs));
+    }
+    unwrapVariables(inputs);
+    // TODO: intepreter needs to accept moved inputs
+    // and delete incrementally
+    tensor_list outputs;
+    InterpreterState(f).runOneStage(std::move(inputs), outputs);
+    wrapTensors(outputs);
+    return outputs;
+  }
+private:
+  // inplace to avoid allocations
+  void unwrapVariables(tensor_list & list) {
+    for(auto & v : list) {
+      v = v.defined() ? autograd::Variable(v).data() : at::Tensor();
+    }
+  }
+  // inplace to avoid allocations
+  void wrapTensors(tensor_list & list) {
+    for(auto & v : list) {
+      v = autograd::make_variable(v);
+    }
+  }
+  // TODO: both captureInputs/captureOutputs have to iterate over the entire
+  // capture list because the list is not sorted with inputs before outputs.
```

Comment:
That is an implementation detail not reflected in the gradients structure. We should update Gradients to actually reflect that structure otherwise relying on it will be fragile.